flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/source/assigners/MySqlSnapshotSplitAssigner.java

```diff
@@ -180,27 +191,60 @@ private void captureNewlyAddedTables() {
         }
     }
 
+    private void asynchronouslySplitIfNeed() {
+        if (!remainingTables.isEmpty()) {
+            if (executor == null) {
+                ThreadFactory threadFactory =
+                        new ThreadFactoryBuilder().setNameFormat("snapshot-split").build();
+                this.executor = Executors.newSingleThreadExecutor(threadFactory);
+            }
+
+            executor.submit(
+                    () -> {
+                        Iterator<TableId> iterator = remainingTables.iterator();
+                        while (iterator.hasNext()) {
+                            TableId nextTable = iterator.next();
+                            // split the given table into chunks (snapshot splits)
+                            Collection<MySqlSnapshotSplit> splits =
+                                    chunkSplitter.generateSplits(nextTable);
+                            synchronized (lock) {
+                                remainingSplits.addAll(splits);
+                                iterator.remove();
+                                lock.notify();
+                            }
+                        }
+                    });
+        }
+    }
+
     @Override
     public Optional<MySqlSplit> getNext() {
         if (!remainingSplits.isEmpty()) {
             // return remaining splits firstly
             Iterator<MySqlSnapshotSplit> iterator = remainingSplits.iterator();
             MySqlSnapshotSplit split = iterator.next();
-            iterator.remove();
+            remainingSplits.remove(split);
             assignedSplits.put(split.splitId(), split);
+            addAlreadyProcessedTablesIfNotExists(split.getTableId());
             return Optional.of(split);
-        } else {
-            // it's turn for next table
-            TableId nextTable = remainingTables.pollFirst();
-            if (nextTable != null) {
-                // split the given table into chunks (snapshot splits)
-                Collection<MySqlSnapshotSplit> splits = chunkSplitter.generateSplits(nextTable);
-                remainingSplits.addAll(splits);
-                alreadyProcessedTables.add(nextTable);
-                return getNext();
-            } else {
-                return Optional.empty();
+        } else if (!remainingTables.isEmpty()) {
```

Comment:
There's a race condition that `remainingTables` is empty but the asynchronous thread just added the snapshot splits of last table to `remainingSplits`, and then the last table will be missed.